### ISSUE TITLE
Authentication Issues using MI with State Store Component: Azure Cosmos DB

### ISSUE THREAD

**Issue by faheemgani** [dapr/components-contrib#2850]
We recently switched over from using a master key to using managed identities to authenticate with Cosmos.

I gave the user assigned identity the following role assignments:

- Cosmos DB Built-in Data Contributor (00000000-0000-0000-0000-000000000002)
- Cosmos DB Account Reader Role
- Cosmos DB Operator Role

These roles are all scoped directly to the Cosmos Account Itself.

I have the following parameters configured in the metadata against the state store component:

- url
- database
- collection
- actorStateStore
- azureClientId (client id of the user assigned managed identity)

When my services try and load the components, they successfully load all other components (pubsub, keyvault etc.) except for the state store component. All other components use managed identities to authenticate too.

This is the error message we are seeing:

time="2023-05-22T12:13:36.29728277Z" level=fatal msg="process component statestore error: [INIT_COMPONENT_FAILURE]: initialization error occurred for statestore (state.azure.cosmosdb/v1): ChainedTokenCredential authentication failed\nGET http://localhost:42356/msi/token\n--------------------------------------------------------------------------------\nRESPONSE 400 Bad Request\n--------------------------------------------------------------------------------\n{\n  \"statusCode\": 400,\n  \"message\": \"No User Assigned or Delegated Managed Identity found for specified ClientId/ResourceId/PrincipalId.\",\n  \"correlationId\": \"xxx\"\n}\n--------------------------------------------------------------------------------\n" app_id=xxx instance=xxx scope=dapr.runtime type=log ver=1.10.4-msft-2

**Comment by ItalyPaleAle** [comment]
@faheemgani can you please confirm where your application is hosted? Azure AKS? Container Apps?

**Comment by faheemgani** [comment]
@ItalyPaleAle We are hosting this in Azure Container Apps

**Comment by berndverst** [comment]
@faheemgani can you confirm the user assigned identity was attached to the container app? This action itself should trigger a restart of your container app.

The error indicates that the identity you specified in `azureClientId` is not actually attached to the Container App.

Instructions for doing so can be found here:

https://learn.microsoft.com/en-us/azure/container-apps/managed-identity?tabs=portal%2Cdotnet#add-a-user-assigned-identity

Please verify that the user assigned identity in question (the one specified in `azureClientId` is actually attached to this particular container app and then try again. Note that you will need to attach the same identity to every single container app which loads this component (or you may need to scope the component to ensure it gets loaded by fewer container apps).

**Comment by faheemgani** [comment]
Hi @berndverst Thank you for your reply. Originally we were trying this using a generic MI that wasn't explicitly 'attached' to the containerapp, we were sharing this across all the containerapps. 

After your suggestion we have programmatically created individual user assigned managed identities for each of our services and we now see this error when trying to load the statestore component:

`time="2023-05-24T03:47:55.982226014Z" level=fatal msg="process component statestore error: [INIT_COMPONENT_FAILURE]: initialization error occurred for statestore (state.azure.cosmosdb/v1): ChainedTokenCredential authentication failed\nGET http://localhost:42356/msi/token\n--------------------------------------------------------------------------------\nRESPONSE 400 Bad Request\n--------------------------------------------------------------------------------\n{\n  \"statusCode\": 400,\n  \"message\": \"Unable to load the proper Managed Identity.\",\n  \"correlationId\": \"f5744ae4-1934-4949-9e5a-3428df2d11a5\"\n}\n--------------------------------------------------------------------------------\n" app_id=sbx-xxx-xxx-api instance=sbx-xxx-xxx-api--t1ul3qe-5c48565cc6-qp8m9 scope=dapr.runtime type=log ver=1.10.4-msft-2`

You can see the user assigned identity is attached to the container app:

![image](https://github.com/dapr/components-contrib/assets/75013259/e482be58-8410-4dbe-8f56-05c3d527bec0)

Here is the managed identity itself - note the last part of the object ID:

![image](https://github.com/dapr/components-contrib/assets/75013259/3da2c7a4-53ad-4594-890d-484d0c3ffa35)

You can see that the managed identity has the correct Cosmos Role Assignment:

![image](https://github.com/dapr/components-contrib/assets/75013259/4039f0a1-8b04-4c3d-8151-42f6e91e4605)

Here is a look at the Dapr Components:

secretstore (works as expected)

![image](https://github.com/dapr/components-contrib/assets/75013259/9cc18537-987d-443c-ab53-b861c9475819)

pubsub (works as expected)

![image](https://github.com/dapr/components-contrib/assets/75013259/c481d508-43a7-4b56-809f-2987ff407a50)

statestore (doesn't work when using MI)

![image](https://github.com/dapr/components-contrib/assets/75013259/1166e76c-d438-4f4e-844b-e3d5d6262a2c)

The other components load okay using the same managed identity (through the same service):

`time="2023-05-24T03:53:00.043526029Z" level=info msg="component loaded. name: pubsub, type: pubsub.azure.servicebus/v1" app_id=sbx-xxx-xxx-api instance=sbx-xxx-xxx-api--t1ul3qe-5c48565cc6-qp8m9 scope=dapr.runtime type=log ver=1.10.4-msft-2`

`time="2023-05-24T03:58:10.97658021Z" level=info msg="component loaded. name: azurekeyvault, type: secretstores.azure.keyvault/v1" app_id=sbx-xxx-xxx-api instance=sbx-xxx-xxx-api--t1ul3qe-5c48565cc6-qp8m9 scope=dapr.runtime type=log ver=1.10.4-msft-2`

**Comment by berndverst** [comment]
Did you manage to resolve this @faheemgani ?

**Comment by faheemgani** [comment]
Hi @berndverst Thank you for your reply. Yes this can be closed now - it has been a bit of a pain to sort, I think we have been experiencing race conditions when using multiple managed identities. 

We tried using two managed identities in each one of our microservices, so that we can have granular control over the permissions assigned to the resources - the goal was to have a unique identity for each one of our microservices and we were using the 'AZURE_CLIENT_ID' env var within the service so that the assigned identity can be picked up and used by the service (this is for authentication to cosmos, service bus etc using the .NET clients and not through Dapr).

The other identity attached to the container app was being shared across all services and the idea was to have this exclusively for the Dapr components. This approach did work for the Dapr components and we stopped seeing the error we were getting during the Component Initialization phase - however - our microservices were inconsistently failing when trying to obtain a token from the MSI_ENDPOINT as I think it was trying to obtain a token for the incorrect identity (as there were two attached to the app) and not the one specified in AZURE_CLIENT_ID env var.

In the end we have decided to have a single identity that we are using - this identity is shared between all the Dapr Components and Container Apps within a specific Container App Environment. This feels a little less secure but is easier to manage permissions and gets us past these issues that we were experiencing.

Thanks for your help. 👍
